### PR TITLE
feat: Read currency pair providers and deviations from on chain params

### PR DESCRIPTION
**Pull request by rbajollari**
## Description

- Adds `chain-config` flag to use currency pair providers and currency deviations from Ojo's oracle params instead of the config file
- Every time the param cache is updated, if the currency pair providers and currency deviations have been changed the price feeder will use the updated ones if it was started using the `chain-config` flag (param cache is updated every 200 blocks or after param update event)

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: cmd/price-feeder.go [hunk 1]
```diff
@@ -61,6 +62,11 @@ func init() {
 	rootCmd.PersistentFlags().String(flagLogLevel, zerolog.InfoLevel.String(), "logging level")
 	rootCmd.PersistentFlags().String(flagLogFormat, logLevelText, "logging format; must be either json or text")
 	rootCmd.PersistentFlags().Bool(flagSkipProviderCheck, false, "skip the coingecko API provider check")
+	rootCmd.PersistentFlags().Bool(
+		flagChainConfig,
+		false,
```
Comment (zarazan) [inline cmd/price-feeder.go:67]: What do you think about making it on by default? If you want to override the on chain values you have to add the flag.

File: oracle/oracle.go [hunk 2]
```diff
@@ -490,6 +541,79 @@ func (o *Oracle) checkAcceptList(params oracletypes.Params) {
...
+		}
+	}
+
+	return nil
+}
+
+// createPairProvidersFromCurrencyPairProvidersList will create the pair providers
+// map used by the price feeder Oracle from a CurrencyPairProvidersList defined by
+// Ojo's oracle module.
+func createPairProvidersFromCurrencyPairProvidersList(
```
Comment (zarazan) [inline oracle/oracle.go:563]: nit: This oracle file is pretty huge and these two functions are not methods of the Oracle struct.I'd prefer them in their own file or another one with related functions.

**Review by zarazan — APPROVED** [review]
Comments are mostly cosmetic. Looks good, very exciting!

File: cmd/price-feeder.go (hunk 1 above)
Comment (rbajollari) [reply cmd/price-feeder.go:67]: I think its better actually yeah. I reversed it to be called `config-currency-providers` which when true uses the config file and is false by default.

File: oracle/oracle.go [hunk 3]
```diff
@@ -82,7 +83,8 @@ func New(
 	providerTimeout time.Duration,
 	deviations map[string]sdk.Dec,
 	endpoints map[types.ProviderName]provider.Endpoint,
-) *Oracle {
+	chainConfig bool,
+) (*Oracle, error) {
```
[the remark this answers is not in the document]
Comment (rbajollari) [reply oracle/oracle.go]: Good catch. I think an earlier implementation I had of this introduced an error that could be returned from this method and I forgot to remove it once I updated that. Changed it back.

File: oracle/oracle.go (hunk 2 above)
Comment (rbajollari) [reply oracle/oracle.go]: Moved to utils👍

**Pull request merged**